ci/remote-watch.py

```diff
@@ -0,0 +1,247 @@
+
+It watches the remote branch for changes, killing the given PID when it detects
+that the remote branch no longer points to the local commit.
+
+If the commit message contains a line saying "CI_KEEP_ALIVE", then killing
+will not occur until the branch is deleted from the remote.
+
+If no PID is given, then the entire process group of this process is killed.
+"""
+
+# Prefer to keep this file Python 2-compatible so that it can easily run early
+# in the CI process on any system.
+
+import errno
+import logging
+import os
+import re
+import signal
+import subprocess
+import sys
+import time
+
+from timeit import default_timer  # monotonic timer
+
+logger = logging.getLogger(__name__)
+
+
+def git(*args, **kwargs):
+    capture = kwargs.pop("capture", True)
+    assert len(kwargs) == 0, "Unexpected kwargs: {}".format(kwargs)
+    cmdline = ["git"] + list(args)
+    if capture:
+        result = subprocess.check_output(cmdline).decode("utf-8").rstrip()
+    else:
+        result = subprocess.check_call(cmdline)
+    return result
+
+
+def get_current_ci():
+    result = None
+    if "GITHUB_WORKFLOW" in os.environ:
+        result = "GitHub"
+    elif "TRAVIS" in os.environ:
+        result = "Travis"
+    return result
+
+
+def get_event_name():
+    result = None
+    ci = get_current_ci()
+    if ci.startswith("GitHub"):
+        result = os.environ["GITHUB_EVENT_NAME"]
+    elif ci.startswith("Travis"):
+        result = os.environ["TRAVIS_EVENT_TYPE"]
+    return result
+
+
+def get_repo_slug():
+    ci = get_current_ci()
+    result = None
+    if ci.startswith("GitHub"):
+        result = os.environ["GITHUB_REPOSITORY"]
+    elif ci.startswith("Travis"):
+        result = os.environ["TRAVIS_REPO_SLUG"]
+    return result
+
+
+def git_branch_info():
+    ref = None
+    try:
+        head = git("symbolic-ref", "-q", "HEAD")
+        ref = git("for-each-ref", "--format=%(upstream:short)", head)
+    except subprocess.CalledProcessError:
+        pass
+    remote = None
+    expected_sha = None
+    if ref:
+        (remote, ref) = ref.split("/", 1)
+        ref = "refs/heads/" + ref
+    else:
+        remote = git_remotes()[0]
+        ref = os.getenv("TRAVIS_PULL_REQUEST_BRANCH")
+        if ref:
+            TRAVIS_PULL_REQUEST = os.environ["TRAVIS_PULL_REQUEST"]
+            if os.getenv("TRAVIS_EVENT_TYPE") == "pull_request":
+                ref = "refs/pull/{}/merge".format(TRAVIS_PULL_REQUEST)
+            else:
+                ref = "refs/heads/{}".format(TRAVIS_PULL_REQUEST)
+            expected_sha = os.getenv("TRAVIS_COMMIT")
+        else:
+            ref = os.getenv("CI_REF")
+    if not remote:
+        raise ValueError("Invalid remote: {!r}".format(remote))
+    if not ref:
+        raise ValueError("Invalid ref: {!r}".format(ref))
+    expected_sha = git("rev-parse", "--verify", "HEAD")
+    return (ref, remote, expected_sha)
+
+
+def terminate_parent_or_process_group(sleeps):
```

Comment:
use psutil?

Reply:
That wouldn't help with anything here would it?

Reply:
My impression was that `psutil` can perform cross platform process killing. Can we use psutil and simplify this 40 lines function into few psutil calls?

Reply:
In addition, because we are just killing the whole job (which in turn will shutdown the CI VM), we don't need to gracefully shutdown processes.

Reply:
https://psutil.readthedocs.io/en/latest/#psutil.wait_procs

Reply:
The graceful shutdown I can change, but `psutil` doesn't really help with that, I can just kill everything without it. (I agree the graceful shutdown isn't quite relevant right now, but it does prevent programs from getting a chance to do useful things like uploading logs down the road, which we'd been thinking about.)

Reply:
Using psutil helps get rid of `is_posix` and `sys.platform == "win32"` and all these special cases.

Reply:
For this script, we just need to kill everything. I don't see the utility in `sleeps` and sleep for some duration before killing.